File: develop/lab-5-sqs-sns/sqsSNSPythonLab/sqs_sns_test.py
# Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights
# reserved.

import unittest
import warnings
import time
import sys
import utils
import sns_publisher
import sqs_consumer
import boto3

QUEUE_NAME = "MySQSQueue_A"
QUEUE_ATTR_NAME = "ApproximateNumberOfMessages"

SLEEP = 10


class SqsSnsTest(unittest.TestCase):

    def test(self):
        warnings.simplefilter("ignore", ResourceWarning)
        try:
            initialNumMessages = self.get_num_of_messages()

            sns_publisher.main()
            numAfterSnsPublisher = self.get_num_of_messages()

            thread1 = sqs_consumer.main()
            thread1.join()
            numAfterSQSConsumer = self.get_num_of_messages()

            print(
                "SqsSnsTest: initialNumMessages: {0}; numAfterSnsPublisher: {1}; numAfterSQSConsumer {2}".format(
                    initialNumMessages,
                    numAfterSnsPublisher,
                    numAfterSQSConsumer))

            if ((numAfterSnsPublisher <= initialNumMessages)
                    or (initialNumMessages != numAfterSQSConsumer)):
                self.fail(
                    "SqsSnsTest failed. Number of messages in queue not as expected.")
        except Exception as err:
            print("Error Message {0}".format(err))
            sys.exit(1)

    def get_queue(self, sqsQueueName=QUEUE_NAME):
        # TODO: Get the SQS queue using the SQS resource object and
        # QUEUE_NAME
        queue = None
        try:
            sqs = boto3.resource('sqs')
            queue = sqs.get_queue_by_name(QueueName=sqsQueueName)
        except Exception as err:
            print("Error Message {0}".format(err))
        return queue

    def get_num_of_messages(self, sqsQueueName=QUEUE_NAME):
        numMessages = 0
        try:
            print("Thread sleeping for a few seconds")
            time.sleep(SLEEP)
            print("Thread running.")
            queue = self.get_queue(sqsQueueName)
            if queue:
                attribs = queue.attributes
                numMessages = int(attribs.get(QUEUE_ATTR_NAME))
        except Exception as err:
            print("Error Message {0}".format(err))
        return numMessages

if __name__ == '__main__':
    unittest.main()

File: develop/lab-5-sqs-sns/sqsSNSJavaLab/src/test/java/com/amazonaws/lab/SqsSnsTest.java
package com.amazonaws.lab;
// Copyright 2017 Amazon Web Services, Inc. or its affiliates. All rights reserved.

import static org.junit.Assert.fail;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.junit.Test;

import com.amazonaws.services.sqs.AmazonSQSClient;
import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
import com.amazonaws.services.sqs.model.GetQueueAttributesResult;
import com.amazonaws.services.sqs.model.GetQueueUrlResult;

public class SqsSnsTest {

	public static final String QUEUE_NAME = SQSConsumer.QUEUE_NAME;
	private static final String QUEUE_ATTR_NAME = "ApproximateNumberOfMessages";
	private static final int SLEEP = 10000;

	private static String queueUrl = null;

	private AmazonSQSClient sqsClient = null;

	@Test
	public void test() throws Exception {
		try {
			init();
			queueUrl = getQueueUrl();

			int initialNumMessages = getNumberOfMessages();

			SNSPublisher.main(new String[0]);

			int numAfterSnsPublisher = getNumberOfMessages();

			SQSConsumer.main(new String[0]);

			int numAfterSQSConsumer = getNumberOfMessages();

			String format = "SqsSnsTest: initialNumMessages: %d; numAfterSnsPublisher: %d; numAfterSQSConsumer %d%n";
			String msg = String.format(format, initialNumMessages, numAfterSnsPublisher, numAfterSQSConsumer);

			System.out.println(msg);

			if ((numAfterSnsPublisher <= initialNumMessages) || (initialNumMessages != numAfterSQSConsumer)) {
				fail("SqsSnsTest failed. Number of messages in queue not as expected.");
			}
		} catch (Exception e) {
			e.printStackTrace();
			throw e;
		}

		System.out.println("SqsSnsTest passed.");
	}

	private void init() throws Exception {
		sqsClient = (AmazonSQSClient) AmazonSQSClientBuilder.standard().build();
	}

	private String getQueueUrl() {
		GetQueueUrlResult queueUrlResult = sqsClient.getQueueUrl(QUEUE_NAME);
		String queueUrl = queueUrlResult.getQueueUrl();
		return queueUrl;
	}

	private int getNumberOfMessages() {
		System.out.printf("SqsSnsTest Thread sleeping for %d milliseconds...", SLEEP);
		try {
			Thread.sleep(SLEEP);
		} catch (InterruptedException ie) {
		}
		System.out.println("SqsSnsTest Thread running.");

		int numMessages = 0;
		List<String> attributeNames = new ArrayList<String>();
		attributeNames.add(QUEUE_ATTR_NAME);
		GetQueueAttributesResult attrResult = sqsClient.getQueueAttributes(queueUrl, attributeNames);
		Map<String, String> attrValueMap = attrResult.getAttributes();
		String attrValue = attrValueMap.get(QUEUE_ATTR_NAME);

		if (attrValue != null) {
			numMessages = Integer.parseInt(attrValue);
		}
		return numMessages;
	}
}